Instead of global data store it in an OPENSSL_CTX

mattcaswell opened the pull request:
This is built on top of the commits in #8728 (Make the EVP API available from within the FIPS module) and addresses the current issues in that PR. Not sure whether to replace that PR with this one or keep it separate.

The new stuff is in the last two commits. Essentially all global data in the core and evp code is moved into the OPENSSL_CTX instead. This was always the intention anyway (see #8760).

A new function "openssl_ctx_run_once" is introduced which is similar to RUN_ONCE except initialisation code is run once per OPENSSL_CTX. We also add a new function "openssl_ctx_onfree" which is similar to OPENSSL_atexit() except it runs when the OPENSSL_CTX is freed.

The ex_data global data is a bit of a special case and needs some separate handling since we can't store that in the normal way in the OPENSSL_CTX (because ex_data is used in the implementation of OPENSSL_CTX).

Having made all these changes it is then possible to share all the same initialisation and de-initialisation code between the FIPS provider and the default provider. We don't need to treat FIPS any differently. Everything is initialised "on demand" as it is normally.

WIP because of the dependency on #8728 and because there are various updates required to the internal documentation that I haven't looked at yet.

paulidale commented:
Just wondering (having not thought this properly through): would it be reasonable to register the "openssl_ctx_run_once" calls like the "openssl_ctx_onfree" ones and make them automatic?

mattcaswell commented:
> Just wondering (having not thought this properly through): would it be reasonable to register the "openssl_ctx_run_once" calls like the "openssl_ctx_onfree" ones and make them automatic?

Where would you register them? As a hard coded list perhaps? There are pros and cons:

pro: you don't have to check all the time whether you've done the initialisation yet
con: initialisation is no longer "on demand" so you might end up allocating lots of resources that you don't end up using

mattcaswell commented:
fixup up pushed to address "make update" issues.

mattcaswell commented:
Another fixup to address a mem leak.

paulidale commented:
> con: initialisation is no longer "on demand" so you might end up allocating lots of resources that you don't end up using

Even if registered before time (a fixed list is a good place), they'd only be called on context creation.  Does that necesarily imply a cascade of other initialisers being called independent of what gets used by the applications??

levitte commented:
Question: did you *have* to put ex_data stuff in the global context for some reason, or did you just take "everything" in #8760 more literally than I meant it?

My thoughts, that I may or may not have shared, is that the internal ex_data stuff would remain global (note: global to the module that includes the ex_data object file)

levitte commented:
> Not sure whether to replace that PR with this one or keep it separate.

I think this should be an entirely separate PR.  There's nothing in this PR that should be inherently dependant on #8728, and having the two mixed in one PR makes it harder to review this one.

mattcaswell commented:
> Question: did you have to put ex_data class data in the global context for some reason, or did you just take "everything" in #8760 more literally than I meant it?

Yes. This solves the problem of the ex_data_lock. By tying it to the lifetime of the OPENSSL_CTX we avoid all the issues we were having with freeing it on the provider side. I see no reason to treat that global data differently to any other global data. In fact its kind of weird not to. This makes things conceptually much simpler.

levitte commented:
> This makes things conceptually much simpler.

I disagree, especially on a conceptual level.  See it like this, the `OSSL_EX_DATA_GLOBAL` is *class data* (not class *instance* data), and with the changes made here, you have basically made it so that you define the *class data* in terms of instances of that class.  Conceptually speaking, that's not simple, not one bit.

mattcaswell commented:
> I disagree, especially on a conceptual level. See it like this, the OSSL_EX_DATA_GLOBAL is class data (not class instance data), and with the changes made here, you have basically made it so that you define the class data in terms of instances of that class. Conceptually speaking, that's not simple, not one bit.

I think you are overcomplicating it. Think of OPENSSL_CTX as a scope. The OSSL_EX_DATA_GLOBAL is class data for all classes within that scope. The OPENSSL_CTX class of objects is slightly different to other classes in that there is only ever one instance within a given scope. I don't think this is a problem.

levitte commented on crypto/context.c:
```diff
@@ -117,3 +135,56 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
+    CRYPTO_THREAD_read_lock(ctx->lock);
+    done = ctx->run_once_done[idx];
+    if (done)
+        ret = ctx->run_once_ret[idx];
```
You don't really need to get the stored return value within guard of the lock...  If `done` is true, the stored return value is constant, period.  If not, this implementation can't be right.

levitte commented on include/openssl/crypto.h:
```diff
@@ -84,6 +84,7 @@ int CRYPTO_atomic_add(int *val, int amount, int *ret, CRYPTO_RWLOCK *lock);
 struct crypto_ex_data_st {
+    OPENSSL_CTX *ctx;
     STACK_OF(void) *sk;
 };
```
Will people be very upset of we make `CRYPTO_EX_DATA` opaque?

mattcaswell replied:
We're supposed to be trying not to break stuff wherever possible. I think this could be a case where we could break things so I think it is best to avoid it.

levitte replied:
Okie...

mattcaswell commented:
I updated this to remove the dependency on #8728.

levitte commented:
> I updated this to remove the dependency on #8728.

Thank you!

levitte commented on crypto/context.c:
```diff
@@ -92,19 +174,88 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
+            return NULL;
+        ctx = default_context;
+    }
+#endif
```
I'm not sure that we need to make this explicitly non-FIPS

mattcaswell replied:
Its dead code in FIPS since we will never have a default context on the provider side there so I'd prefer to compile out the RUN_ONCE path altogether.

richsalz replied:
If it's dead code it's kinda harmless and I think you should consider "minimize ifdefs" as a goal as well, even if you come to the same conclusion.

mattcaswell replied:
Well let me put it another way. It *should* be dead code in the FIPS provider. In the FIPS provider `ctx` should never be NULL on entry to this function. If for some reason a bug means that we enter this function and it *is* NULL then with this ifdef there we will fail immediately. Without it there we will continue on, it will automatically allocate a ctx completely disconnected from the ctx it is supposed to be using and weird stuff will happen - probably failing in some completely unrelated location - making debugging very hard.

levitte replied:
I kinda suspected that's what you wanted.  Please add commentary on the subject if you haven't already (I haven't looked for that...)

mattcaswell replied:
Comment added

levitte commented on crypto/evp/evp_fetch.c:
```diff
@@ -68,9 +57,9 @@ struct method_data_st {
+static void *alloc_tmp_method_store(OPENSSL_CTX *ctx)
 {
-    return ossl_method_store_new();
+    return ossl_method_store_new(ctx);
```
It took me a bit to see this myself, so I think we need to comment that `ctx` doesn't mean the allocated store ends up getting stored in in the library context but that `ctx` allows access to underlying property data needed for computation.

mattcaswell replied:
Done

levitte commented on crypto/provider_core.c:
```diff
@@ -134,23 +133,24 @@ static const OPENSSL_CTX_METHOD provider_store_method = {
-static CRYPTO_ONCE provider_store_init_flag = CRYPTO_ONCE_STATIC_INIT;
-DEFINE_RUN_ONCE_STATIC(do_provider_store_init)
+static openssl_ctx_run_once_fn do_provider_store_init;
+int do_provider_store_init(OPENSSL_CTX *ctx)
```
I know that you don't *have* to, but I'd rather `static` git repeated on this line.

mattcaswell replied:
Thanks - that was an over sight.

levitte commented:
The reason you're getting build errors on Windows is the new inclusion of `"internal/cryptlib.h"` in `include/internal/property.h`, which is included by `test/property_test.c`.  The simple solution is this:

``` diff
: ; git diff
diff --git a/test/build.info b/test/build.info
index ded3bd770a..2800c71d3e 100644
--- a/test/build.info
+++ b/test/build.info
@@ -503,7 +503,7 @@ IF[{- !$disabled{tests} -}]
     DEPEND[wpackettest]=../libcrypto ../libssl.a libtestutil.a
 
     SOURCE[property_test]=property_test.c
-    INCLUDE[property_test]=../include ../apps/include
+    INCLUDE[property_test]=.. ../include ../apps/include
     DEPEND[property_test]=../libcrypto.a libtestutil.a
 
     SOURCE[ctype_internal_test]=ctype_internal_test.c
```

levitte commented:
These manuals need an update:

- `doc/internal/man3/ossl_method_construct.pod`
- `doc/internal/man3/OSSL_METHOD_STORE.pod`
- `doc/internal/man3/openssl_ctx_get_data.pod`

mattcaswell commented:
> Just wondering (having not thought this properly through): would it be reasonable to register the "openssl_ctx_run_once" calls like the "openssl_ctx_onfree" ones and make them automatic?

After a bit of "Doh!" moment I realised that we could do this with all of the openssl_ctx_init_index calls - and suddenly you don't need some of the run_once calls at all because a number of them were only there to initialise the index. Furthermore, I realised I wasn't using OPENSSL_CTX quite right and I could move even more initialisation code into the OPENSSL_CTX_METHOD "new" functions. This significantly simplifies things. Most code can just set up an OPENSSL_CTX_METHOD and it all just initialises itself when needed and frees itself when the OPENSSL_CTX_METHOD is freed. No need to call openssl_ctx_run_once or openssl_ctx_onfree at all.

There is still one spot where openssl_ctx_run_once is required. There are no places where openssl_ctx_onfree is now needed, although I have left the implementation there since I suspect it will be needed as we roll this out to more places in the code.

New commits pushed addressing the feedback and making the changes described above. Taking this out of WIP.

mattcaswell commented:
>  The simple solution is this

Thanks for the hint. Should be sorted now. Also I updated the internal manuals as requested.

levitte commented on crypto/context.c:
```diff
@@ -15,6 +15,19 @@ struct openssl_ctx_onfree_list_st {
+    &provider_store_method,
+    &property_defns_method,
+    &property_string_data_method
+};
```
Please no.  This is not a good idea, at all.  If nothing else, then that comment should indicate that this is a maintenance problem waiting to happen.  There's really no reason why code that initiate an index shouldn't know their corresponding `OPENSSL_CTX_METHOD`.

mattcaswell replied:
An alternative I considered was to pass the method every time we call openssl_ctx_get_data. I thought this approach was nicer but could go that way if you prefer.

levitte replied:
The idea was that each sub-system using this would use a pattern like this:

``` C
static int init_my_store(OPENSSL_CTX *libctx)
{
    return openssl_ctx_init_index(libctx, MY_INDEX, &my_method);
}

static blah *get_my_store(OPENSSL_CTX *libctx)
{
    if (!openssl_ctx_run_once(libctx, MY_INDEX, init_my_store))
       return NULL;

    /* At this point, the index is correct and its method initiated */
    return openssl_ctx_get_data(libctx, MY_INDEX);
}
```

Anything else would be an invalid implementation.

mattcaswell replied:
Yes, I know that was the idea. And that's what I had originally. But then I ended up with loads of openssl_ctx_run_once calls and init functions that *only* initialised the index. There was a lot of extra code that essentially did very little. If I got rid of the need for that everything suddenly became a lot simpler.

levitte replied:
... and that leads to a managment problem.  I find that *more* displeasing

levitte commented on crypto/context.c:
```diff
@@ -158,8 +177,11 @@ int openssl_ctx_init_index(OPENSSL_CTX *ctx, int static_index,
+    idx = crypto_get_ex_new_index_ex(ctx, CRYPTO_EX_INDEX_OPENSSL_CTX, 0,
+                                     meth == NULL
+                                     ? (void *)methods[static_index]
+                                     : (void *)meth,
```
I'd rather that `meth == NULL` be an error.  Considering it's all internal, I think it's justifiable if that leads to a crash, i.e. we should fix the issue before release.

mattcaswell replied:
The *only* reason "meth == NULL" is here at all is to support the tests. Its the only caller where meth != NULL. As noted above an alternative is pass the method in the openssl_ctx_get_data call.

levitte replied:
Ah, not originally, but with the latest changes.

levitte commented on crypto/context.c:
```diff
@@ -172,16 +194,24 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return NULL;
+    }
+
```
Ah, I see what you're trying to do.  I think you're trying to make the code too smart.  If the index is -1, this should simply be an error.  (I actually let that mean that `CRYPTO_alloc_ex_data` crash, but returning NULL is fine too)

mattcaswell replied:
I think this avoids *lots* of complexity. It actually means in most cases you don't need run once init code *at all*.

levitte replied:
Have you noticed the times when outside people have asked "if I want to add a sub-system, what things must I touch", and it turns out that they must go touch a number of things, and it's way too easy to miss something?  This recreates exactly that situation.  I was trying to make the slot allocation in the library context fairly self contained within each sub-system that does it, and the code changes I see now go exactly in the opposite direction.

mattcaswell replied:
If you don't like that then we can go with the "pass the method as part of the openssl_ctx_get_data" call idea. Actually we could make a macro for each file we do that (if there are multiple of them), which knows the index and method for that file already. So something like

````
#define FOO_get_openssl_ctx_data(ctx)      openssl_ctx_get_data(ctx, FOO_INDEX, foo_ctx_method)
````

levitte replied:
That makes me cringe a little less.  At least, that preserves the more decentralized creation of the index.  So, uhmm, yes please (not sure I see the necessity for the macro, but [shrug])

mattcaswell commented:
Updated to address @levitte's feedback.

levitte commented:
So, what about https://github.com/openssl/openssl/pull/8857#discussion_r280357779?

levitte commented:
(other than that, I'm not reasonably happy with this PR)

mattcaswell commented:
> So, what about #8857 (comment)?

Responded to

mattcaswell commented:
Pusehd! Thanks.

t8m commented:
> (other than that, I'm not reasonably happy with this PR)
I suppose you meant now instead of not :)

levitte commented:
> > (other than that, I'm not reasonably happy with this PR)
> 
> I suppose you meant now instead of not :)

Uhmmmm... yes!
